Content.Shared/Damage/DamageableSystem.cs

```diff
@@ -0,0 +1,291 @@
+        {
+            var damageGroupDict = new Dictionary<DamageGroupPrototype, int>();
+            foreach (var group in _prototypeManager.EnumeratePrototypes<DamageGroupPrototype>())
+            {
+                var groupDamage = 0;
+                var groupIsSupported = false;
+
+                foreach (var type in group.DamageTypes)
+                {
+                    if (component.DamagePerType.TryGetValue(type, out var damage))
+                    {
+                        groupIsSupported = true;
+                        groupDamage += damage;
+                    }
+
+                }
+
+                // was at least one member of this group actually supported by the container?
+                if (groupIsSupported)
+                {
+                    damageGroupDict.Add(group, groupDamage);
+                }
+            }
+            return damageGroupDict;
+        }
+    }
+
+    public class DamageChangedEvent : EntityEventArgs
+    {
+        /// <summary>
+        ///     This is the complete information about the new damage state.  
+        /// </summary>
+        /// <remarks>
+        ///     Given that nearly every component that cares about a change in the damage, needs to know the
+        ///     current damage values, directly passing this information prevents a lot of duplicate
+        ///     Owner.TryGetComponent() calls. One of the few exceptions is lightbulbs, which just care if ANY damage
+        ///     was taken, not how much.
+        /// </remarks>
+        public DamageableComponent Damageable { get; }
+
+        /// <summary>
+        ///     Has any damage type increased? 
+        /// </summary>
+        /// <remarks>
+        ///     This can still be true even if the overall effect of the damage change was to reduce the total damage.
+        /// </remarks>
+        public bool TookDamage { get; }
+        public DamageChangedEvent(DamageableComponent damageable, bool tookDamage)
+        {
+            Damageable = damageable;
+            TookDamage = tookDamage;
+        }
+    }
+
+    /// <summary>
+    /// Event used to deal or heal damage on a damageable component. Handled by <see cref="DamageableSystem"/>
+    /// </summary>
+    public class TryChangeDamageEvent : EntityEventArgs
+    {
+        /// <summary>
+        /// Input. Damage that is added to the DamageableComponent
+        /// </summary>
+        public DamageData Damage { get; }
+
+        /// <summary>
+        /// Input. Whether to ignore resistances of the damageable component. Healing ignores resistances.
+        /// Defaults to false.
+        /// </summary>
+        public bool IgnoreResistances { get;  }
+
+        /// <summary>
+        /// Output. Did the damage actually change? Maybe it was all resisted, or you healed someone at full health.
+        /// </summary>
+        public bool DidDamageChange { get; set; } = false;
```

Comment:
Currently, there are very few bits of code (ignoring tests) that actually care about whether any damage was changed by the event. Maybe this should just be removed?